sctp.c

```diff
@@ -301,12 +301,11 @@ int janus_sctp_association_setup(janus_sctp_association *sctp) {
 }
 
 void janus_sctp_association_destroy(janus_sctp_association *sctp) {
-	if(sctp == NULL)
+	if(sctp == NULL || !g_atomic_int_compare_and_exchange(&sctp->destroyed, 0, 1))
 		return;
 	usrsctp_deregister_address(sctp);
 	usrsctp_shutdown(sctp->sock, SHUT_RDWR);
 	usrsctp_close(sctp->sock);
-	g_atomic_int_set(&sctp->destroyed, 1);
```

Comment:
I don't know enough at the moment to know whether this change is correct or not, but I *have* observed occasional segfaults related to SCTP transmission directly when I destroy a session on the refcount branch for the previous few days after upgrading to the tip of the branch from f72f223b3852ff76bc77b33d5c27d951d506563d. Might this be the culprit? It does seem to have a different behavior in one weird case (namely, if the association was already destroyed and this gets called again, then it won't try to run the same shutdown code again.)

I can send a core dump if it helps. I haven't reproduced the crash reliably yet -- I'll let you know if I can. Here's the stack trace:

```
>>> bt
#0  0x00007fa43ae2e126 in __memmove_avx_unaligned_erms () at ../sysdeps/x86_64/multiarch/memmove-vec-unaligned-erms.S:525
#1  0x00007fa43bd15ec6 in  () at /lib/x86_64-linux-gnu/libcrypto.so.1.0.0
#2  0x00007fa43bd14ed5 in BIO_write () at /lib/x86_64-linux-gnu/libcrypto.so.1.0.0
#3  0x0000556d578e2602 in janus_dtls_bio_filter_write (bio=0x7fa408008d60, in=0x7fa408013aca "", inl=-93) at dtls-bio.c:133
#4  0x00007fa43bd14ed5 in BIO_write () at /lib/x86_64-linux-gnu/libcrypto.so.1.0.0
#5  0x00007fa43c06a9a2 in  () at /lib/x86_64-linux-gnu/libssl.so.1.0.0
#6  0x0000556d578e1b14 in janus_dtls_send_sctp_data (dtls=0x7fa408006530, buf=<optimized out>, len=<optimized out>) at dtls.c:906
#7  0x0000556d57913d33 in janus_sctp_thread (data=0x7fa4080187c0) at sctp.c:1298
#8  0x00007fa43c52c645 in  () at /lib/x86_64-linux-gnu/libglib-2.0.so.0
#9  0x00007fa43b0967fc in start_thread (arg=0x7fa421de9700) at pthread_create.c:465
#10 0x00007fa43adc3b0f in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
```

Reply:
Can't look in any log as I'm abroad for a conference, but Github marks this as an outdated comment: have you checked if this hasn't been fixed in master?

Reply:
Checked the code and it shouldn't be it, as while we removed the explicit "set destroyed to 1", it's still there in the `g_atomic_int_compare_and_exchange` a few lines before. We changed the code to make `janus_sctp_association_destroy` idempotent, so that if you call it twice you don't actually unref and destroy stuff twice.

Reply:
Ah, did you see my edit? Maybe it's making a race condition more likely than it was before thanks to the reordering.

Reply:
(Sorry that it got marked as outdated, I was commenting on the commit because I was looking through specifically the most recent commits under the hypothesis that my refcount branch used to work reliably and now has the bug, which may or may not be really true.)